TSPL.docc/LanguageGuide/Concurrency.md

```diff
@@ -564,33 +637,192 @@ the explicit parent-child relationships between tasks
 let Swift handle some behaviors like propagating cancellation for you,
 and lets Swift detect some errors at compile time.
 
+[`TaskGroup`]: https://developer.apple.com/documentation/swift/taskgroup
+
+<!-- TR: What's a good example of this kind of error? -->
+
+Here's another version of the code to download photos
+that handles any number of photos:
+
 ```swift
 await withTaskGroup(of: Data.self) { taskGroup in
+    let photoNames = await listPhotos(inGallery: "Summer Vacation")
+    for name in photoNames {
+        taskGroup.addTask {
+            let photo = downloadPhoto(named: name)
+            show(photo)
+        }
+    }
+}
+```
+
+The code above creates a new task group,
+and then creates child tasks of that task group
+to download each photo is the gallery.
+Each task group downloads a single photograph
+and then adds it to the photographs being displayed.
+Because there isn't a specific priority passed
+when calling `addTask(priority:operation:)`,
+these child tasks tasks get the same priority as the task group.
+Swift schedules these tasks,
+running as many of them at the same time as conditions allow.
+
+> Note:
+> If the code to download a photo could throw an error,
+> you would call `withThrowingTaskGroup(of:returning:body:)` instead.
+
+In the code listing above,
+each photo is downloaded and immediately displayed,
+so the task group doesn't return any results.
+For a task group that returns a result,
+you add code that accumulates its result
+inside the closure you pass to `withTaskGroup(of:returning:body:)`.
+
+```
+let photos = await withTaskGroup(of: Data.self) { taskGroup in
     let photoNames = await listPhotos(inGallery: "Summer Vacation")
     for name in photoNames {
         taskGroup.addTask { await downloadPhoto(named: name) }
     }
+
+    var results: [Data] = []
+    for await photo in taskGroup {
+        results.append(photo)
+    }
+    return results
 }
 ```
 
-<!--
-  TODO walk through the example
+Like the previous example,
+this example creates a child task for each photo to download it.
+Unlike the previous example,
+the `for`-`await`-`in` loop waits for the next child task to finish,
+appends the result of that task to the array of results,
+and then continues waiting until all child tasks have finished.
+Finally,
+the task group returns the array of downloaded photos
+as its overall result.
+
```

Comment:
It might be cool to use the photos example to include a demonstration on limiting the number of concurrent tasks spun up by a task group here, kind of like the [chopping tasks](https://developer.apple.com/videos/play/wwdc2023/10170?time=688) in the WWDC video.

Reply:
Do we have general guidance we want to give about how many concurrent tasks a program should be running?  That is, if we show how to set a limit, we should show how to pick a good number to use as that limit.

Reply:
Nothing general that I know of. It sort of depends on the OS, number of cores, what the tasks are doing/waiting on.
I think it would be another one for the "adjust based on profile information" book.

Reply:
Saved in a TODO for the future (428e3414).